activerecord/lib/active_record/core.rb

```diff
@@ -19,11 +19,19 @@ module Core
       # retrieved on both a class and instance level by calling +logger+.
       class_attribute :logger, instance_writer: false
 
-      ##
-      # :singleton-method:
-      #
-      # Specifies the job used to destroy associations in the background
-      class_attribute :destroy_association_async_job, instance_writer: false, instance_predicate: false, default: false
+      class_attribute :_destroy_association_async_job, instance_writer: false, instance_predicate: false, default: "ActiveRecord::DestroyAssociationAsyncJob"
```

Comment:
Now comes the tricky part. This job will not be loaded in production until the first time `destroy_association_async_job` is called, so we need to make sure when eager loading the application we call that method.

I'm not sure this would not trigger the same problem you saw in the first place, but we should not wait for the first use of this job in production.

Reply:
I believe this will triggered by [`check_dependent_options`](https://github.com/rails/rails/blob/c3c747547a85937e67bc62cfb9f466948fb631ad/activerecord/lib/active_record/associations/builder/association.rb#L129) which will run when the association is declared (e.g. `has_many ... dependent: :destroy_async`).  Though that does seem a little fragile.  Perhaps we could add a test to ensure that `destroy_association_async_job` is called?

Reply:
Correct `destroy_association_async_job` will be called (and constant loaded) when `check_dependent_options` is called when the association is declared. I've updated the tests to cover the scenarios where the constant is invalid/unloadable (which raises a NameError with a custom message), and where the constant is nil but `dependent: :destroy_async` is configured (the exception in this case changed from `ActiveRecord::ActiveJobRequiredError` to an `ActiveRecord::ConfigurationError` with a custom message).

> we need to make sure when eager loading the application we call that method.

The Job will be eager-loaded when application's ActiveRecord models are eager-loaded.

> I'm not sure this would not trigger the same problem you saw in the first place

I don't believe it will. This change means that the Job constant (and ActiveJob::Base) will be loaded when the ActiveRecord::Base _child class_ model is loaded. `ActiveRecord::Base` should be fully loaded at that point.